Miscellaneous/Best practices for error handling and crash reporting.md

```diff
@@ -0,0 +1,434 @@
+
+We can expose these details in our `ErrorView` by providing an additional Details button. If you can’t do that in production, it might be a good idea to enable detailed error handling at least in staging builds.
+
+This can simplify finding issues and shorten the feedback loop between the person reproducing the error and you.
+
+
+###### Non-fatals
+
+Firebase [non fatals](https://firebase.google.com/docs/crashlytics/customize-crash-reports?platform=ios#log-excepts) are a great way to get more detailed information about errors that occur in production. This can usually provide critical information for debugging user-reported issues. To utilize and analyze them effectively, we need to have fine fine-grained grouping of errors.
+
+As mentioned in the [Crashlytics guide](https://firebase.google.com/docs/crashlytics/customize-crash-reports?platform=ios):
+
+“Unlike fatal crashes, which are grouped via stack trace analysis, logged errors are
+
+grouped by _domain_ and _code_.”
+
+This means that, to group Swift errors properly, we need to conform them to `CustomNSError`. For example, our `APIError` can conform to `CustomNSError` in the following way: 
+
+```
+extension APIError: CustomNSError {
+    var errorCode: Int { code }
+    static var errorDomain: String { "APIError" }
+}
+```
+
+This will group all API errors with the same status code.
+
+The additional benefit of conforming to `CustomNSError` is that we can provide additional details that will show up in Firebase:
+
+```
+var errorUserInfo: [String: Any] {
+    [
+        "title": response?.title as Any,
+        "detail": response?.detail as Any,
+    ]
+}
+```
+
+## Application crashes
+
+An application crash is usually the worst outcome our app users can experience. And we should do everything in our power to prevent that.
+
+
+### Fail fast
+
+However, avoiding crashes is not something to be done at any cost.
+
+Generally speaking, when we can’t handle some condition in a meaningful way, we should expose this as an error to the user.
+
+But, we are also limited by the programming language and ecosystem we operate in. Sometimes we need more expressivity in the language to represent some concept.
+
+Let’s look at the following example:
+
+```
+class NamePicker {
+    private let names = ["John", "Josh", "Lucy", "Angela"]
+    private var selectedName: String
+
+    init() {
+        selectedName = names.first!
+    }
+}
+```
+
+Generally speaking, force unwrapping is a pattern that we would like to avoid. But in this case, making `init` `throwing` wouldn’t convey the right message and it would pollute the code. It would require all upstream functions to be annotated with throws which would obscure the true value of genuine errors.
+
+Here is another example:
+
+```
+let status = CVDisplayLinkCreateWithActiveCGDisplays(&link)
+guard let link, status == kCVReturnSuccess else {
+    fatalError("Could not create display link. Return status: \(status)")
```

Comment:
Just out of curiosity, what is your stance on assertionFailure, and preconditionFailure usage? Do you even use them?

Reply:
`preconditionFailure` is similar to `fatalError` in its behaviour, but maybe better conveys intention in some scenarios.

`assertionFailure` is interesting one. In new code I try to avoid this as much as possible, because this is introducing undefined behaviour in production builds.

However, for legacy code, I think it can be a useful tool. I am afraid of crashing under certain conditions that should not happen, but might since I don't have full picture. In this case, crashing in debug builds can be useful tool to discover those conditions, while still keeping production code working as is.